FIAT/brezzi_douglas_marini_cube.py

```diff
@@ -0,0 +1,229 @@
+        points = list(map(transform, points))
+
+        phivals = {}
+
+        for o in range(order+1):
+            alphas = mis(2, o)
+            for alpha in alphas:
+                try:
+                    polynomials = self.basis[alpha]
+                except KeyError:
+                    polynomials = diff(self.basis[(0, 0)], *zip(variables, alpha))
+                    self.basis[alpha] = polynomials
+                T = np.zeros((len(polynomials[:, 0]), 2, len(points)))
+                for i in range(len(points)):
+                    subs = {v: points[i][k] for k, v in enumerate(variables[:2])}
+                    for j, f in enumerate(polynomials[:, 0]):
+                        T[j, 0, i] = f.evalf(subs=subs)
+                    for j, f in enumerate(polynomials[:, 1]):
+                        T[j, 1, i] = f.evalf(subs=subs)
+                phivals[alpha] = T
+
+        return phivals
+
+    def entity_dofs(self):
+        """Return the map of topological entities to degrees of
+        freedom for the finite element."""
+        return self.entity_ids
+
+    def entity_closure_dofs(self):
+        """Return the map of topological entities to degrees of
+        freedom on the closure of those entities for the finite element."""
+        return self.entity_closure_ids
+
+    def value_shape(self):
+        return (2,)
+
+    def dmats(self):
+        raise NotImplementedError
+
+    def get_num_members(self, arg):
+        raise NotImplementedError
+
+    def space_dimension(self):
+        return int(len(self.basis[(0, 0)])/2)
+
+
+def e_lambda_1_2d(deg, dx, dy, x_mid, y_mid):
+    EL = tuple([(0, -leg(j, y_mid)*dx[0]) for j in range(deg)] +
+               [(-leg(deg-1, y_mid)*dy[0]*dy[1]/(deg+1), -leg(deg, y_mid)*dx[0])] +
+               [(0, -leg(j, y_mid)*dx[1]) for j in range(deg)] +
+               [(leg(deg-1, y_mid)*dy[0]*dy[1]/(deg+1), -leg(deg, y_mid)*dx[1])] +
+               [(-leg(j, x_mid)*dy[0], 0) for j in range(deg)] +
+               [(-leg(deg, x_mid)*dy[0], -leg(deg-1, x_mid)*dx[0]*dx[1]/(deg+1))] +
+               [(-leg(j, x_mid)*dy[1], 0) for j in range(deg)] +
+               [(-leg(deg, x_mid)*dy[1], leg(deg-1, x_mid)*dx[0]*dx[1]/(deg+1))])
+
+    return EL
+
+
+def f_lambda_1_2d(deg, dx, dy, x_mid, y_mid):
+    FL = []
+    for k in range(2, deg+1):
+        for j in range(k-1):
+            FL += [(0, leg(j, x_mid)*leg(k-2-j, y_mid)*dx[0]*dx[1])]
+            FL += [(leg(k-2-j, x_mid)*leg(j, y_mid)*dy[0]*dy[1], 0)]
+
+    return tuple(FL)
+
+
+class BrezziDouglasMariniCubeEdge(BrezziDouglasMariniCube):
+    def __init__(self, ref_el, degree):
+        if degree < 1:
+            raise Exception("BDMcf_k elements only valid for k >= 1")
+
+        flat_el = flatten_reference_cube(ref_el)
+        dim = flat_el.get_spatial_dimension()
+        if dim != 2:
+            raise Exception("BDMcf_k elements only valid for dimension 2")
```

Comment:
Seems strange to call something a cube element when it's only valid for squares.